01_stratified_classification/nbconverted/lasso_range_gene_optimizers.py

```diff
@@ -0,0 +1,791 @@
+
+
+# In[21]:
+
+
+sns.set_style('ticks')
+
+with sns.plotting_context('notebook', font_scale=1.6):
+    g = sns.relplot(
+        data=perf_coefs_df,
+        x='nz_quantile', y=metric, hue='data_type',
+        hue_order=['train', 'cv', 'test'],
+        marker='o', kind='line', col='optimizer',
+        col_wrap=2, height=5, aspect=1.6,
+        facet_kws={'sharex': False}
+    )
+    g.axes[0].set_xlabel('Bin (increasing number of nonzero coefficients)')
+    g.axes[0].set_xlim((0, perf_coefs_df.nz_quantile.max()))
+    g.axes[1].set_xlabel('Bin (increasing number of nonzero coefficients)')
+    g.axes[1].set_xlim((0, perf_coefs_df.nz_quantile.max()))
+    g.set_ylabels(f'{metric.upper()}')
+    sns.move_legend(g, "center", bbox_to_anchor=[1.045, 0.55], frameon=True)
+    g._legend.set_title('Dataset')
+    new_labels = ['train', 'holdout', 'test']
+    for t, l in zip(g._legend.texts, new_labels):
+        t.set_text(l)
+    g.set_titles('Optimizer: {col_name}')
+     
+    plt.suptitle(
+        f'Number of nonzero coefficients vs. {metric.upper()}, decile binning, {plot_gene}',
+        y=1.0
+    )
+
+plt.tight_layout()
+
+if output_plots:
+    plt.savefig(os.path.join(output_plots_dir, f'{gene}_decile_vs_perf.svg'), bbox_inches='tight')
+
+
+# In[22]:
+
+
+decile_perf_df = (perf_coefs_df
+    [perf_coefs_df.data_type == 'test']
+    .groupby(['optimizer', 'nz_quantile'])
+    .agg(np.mean)
+    .drop(columns=['seed', 'fold', 'lasso_param'])
+)
+
+print(decile_perf_df.shape)
+decile_perf_df.loc['liblinear', :]
+
+
+# In[23]:
+
+
+decile_perf_df.loc['SGD', :]
+
+
+# In[24]:
+
+
+# plot coefficient distributions for this seed/fold
+plot_seed = 42
+plot_fold = 0
+
+
+# In[25]:
+
+
+ll_top_df = []
+
+# get top-performing lasso param for each gene,
+# based on mean performance across seeds/folds
+ll_mean_perf_df = (
+  ll_perf_df[(ll_perf_df.data_type == 'cv') &
+             (ll_perf_df.signal == 'signal')]
+      .groupby(['lasso_param'])
+      .agg(np.mean)
+      .drop(columns=['seed', 'fold'])
+      .rename(columns={'auroc': 'mean_auroc', 'aupr': 'mean_aupr'})
+      .sort_values(by='mean_aupr', ascending=False)
+      .reset_index()
+)
+ll_mean_perf_df.head()
+
+
+# In[26]:
+
+
+sgd_top_df = []
+
+# get top-performing lasso param for each gene,
+# based on mean performance across seeds/folds
+sgd_mean_perf_df = (
+  sgd_perf_df[(sgd_perf_df.data_type == 'cv') &
+              (sgd_perf_df.signal == 'signal')]
+      .groupby(['lasso_param'])
+      .agg(np.mean)
+      .drop(columns=['seed', 'fold'])
+      .rename(columns={'auroc': 'mean_auroc', 'aupr': 'mean_aupr'})
+      .sort_values(by='mean_aupr', ascending=False)
+      .reset_index()
+)
+sgd_mean_perf_df.head()
+
+
+# In[27]:
+
+
+ll_top_lasso_param = ll_mean_perf_df.iloc[0, :].lasso_param
+sgd_top_lasso_param = sgd_mean_perf_df.iloc[0, :].lasso_param
+print(ll_top_lasso_param, sgd_top_lasso_param)
+
+
+# In[28]:
```

Comment:
For readability, the rest of the file should probably be its own subsection about comparing coefficient size, with some explanation of why you're interested in it.